File: src/main/java/co/bitshifted/xapps/backstage/deploy/builders/WindowsDeploymentBuilder.java
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.backstage.deploy.builders;

import co.bitshfted.xapps.zsync.ZsyncMake;
import co.bitshifted.xapps.backstage.content.ContentMapping;
import co.bitshifted.xapps.backstage.deploy.DeploymentConfig;
import co.bitshifted.xapps.backstage.deploy.TargetDeploymentInfo;
import co.bitshifted.xapps.backstage.deploy.XmlProcessor;
import co.bitshifted.xapps.backstage.dto.UpdateDetail;
import co.bitshifted.xapps.backstage.dto.UpdateInformation;
import co.bitshifted.xapps.backstage.exception.DeploymentException;
import co.bitshifted.xapps.backstage.model.CpuArch;
import co.bitshifted.xapps.backstage.model.OS;
import co.bitshifted.xapps.backstage.service.UpdateService;
import co.bitshifted.xapps.backstage.util.BackstageFunctions;
import co.bitshifted.xapps.backstage.util.PackageUtil;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;

import static co.bitshifted.xapps.backstage.BackstageConstants.*;

/**
 * @author Vladimir Djurovic
 */
@Slf4j
public class WindowsDeploymentBuilder extends DeploymentBuilder {


	@Autowired
	private ContentMapping contentMapping;
	@Value("${update.server.baseurl}")
	private String updateServerBaseUrl;

	public WindowsDeploymentBuilder(TargetDeploymentInfo deploymentInfo) {
		super(deploymentInfo);
	}

	@Override
	protected Path prepareDirectoryStructure() throws DeploymentException {
		try {
			var dirName = targetOs.getBrief() + "-" + targetCpuArch.getDisplay();
			var buildDirectory = deploymentWorkDir.resolve(dirName);
			Files.createDirectory(buildDirectory);
			logger().debug("Created Windows build directory {}", buildDirectory);
			var appDirPath = buildDirectory.resolve(config.getAppName());
			Files.createDirectory(appDirPath);
			return appDirPath;
		} catch (IOException ex) {
			log.error("Failed to create build directory");
			throw new DeploymentException(ex);
		}

	}

	@Override
	public void createDeployment() throws DeploymentException {
		try {
			// create windows build directory
			var appDirPath = prepareDirectoryStructure();
			// add modules for updates
			copySyncroModules();
			createJreImage(appDirPath.resolve("jre"));
			copyIcons(appDirPath);
			copySplashScreen(appDirPath);
			copyLauncher(LAUNCHER_FILE_NAME_WIN_64, appDirPath);
			// create launcher configuration
			writeLauncherConfig(appDirPath);
			var updatePkgPath = createUpdatePackage(appDirPath);
			log.debug("Update package path: {}", updatePkgPath.toString());
			prepareForDownload(updatePkgPath);
		} catch (Exception ex) {
			log.error("Failed to create deployment", ex);
			throw new DeploymentException(ex);
		}

	}

	@Override
	protected Path createUpdatePackage(Path appDir) throws DeploymentException {
		try {
			var updateDir = appDir.getParent().resolve("update");
			Files.createDirectories(updateDir);
			log.debug("Created update directory {}", updateDir.toString());
			var jreModules = appDir.resolve("jre/lib/modules");
			var moved = updateDir.resolve("modules");
			Files.move(jreModules, moved);
			var updatePath = PackageUtil.packZipDeterministic(appDir);
			// update package contents
			var updateContents = updateDir.resolve("contents.zip");
			Files.move(updatePath, updateContents);
			log.debug("Created contents update package at {}", updateContents.toString());
			// compress modules file
			var updateModules = PackageUtil.zipSingleFile(moved);
			log.debug("Created update modules file {}", updateModules.toString());
			Files.move(moved, jreModules);
			return updateDir;
		} catch (IOException ex) {
			throw new DeploymentException(ex);
		}
	}

	@Override
	protected Logger logger() {
		return log;
	}
}

File: src/main/kotlin/co/bitshifted/backstage/service/deployment/builders/WindowsDeploymentBuilder.kt
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.backstage.service.deployment.builders

import co.bitshifted.backstage.BackstageConstants
import co.bitshifted.backstage.exception.BackstageException
import co.bitshifted.backstage.exception.ErrorInfo
import co.bitshifted.backstage.util.logger
import co.bitshifted.ignite.common.model.OperatingSystem
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import kotlin.io.path.absolutePathString

class WindowsDeploymentBuilder(val builder: DeploymentBuilder) {

    val logger = logger(this)
    lateinit var classpathDir: Path
    lateinit var modulesDir: Path

    fun build(): Boolean {
        logger.info("Creating Windows deployment in directory {}", builder.windowsDir)
        try {
            createDirectoryStructure()
            builder.copyDependencies(modulesDir, classpathDir, OperatingSystem.WINDOWS)
            builder.copyResources(builder.windowsDir)
            builder.buildJdkImage(builder.windowsDir, modulesDir, OperatingSystem.WINDOWS)
            copyLauncher()
            copyWindowsIcons()
            copySplashScreen()
            logger.info("Successfully created Windows deployment in directory {}", builder.windowsDir)
            return true
        } catch (th: Throwable) {
            throw th
        }
    }

    private fun createDirectoryStructure() {
        classpathDir = Files.createDirectories(
            Paths.get(
                builder.windowsDir.absolutePathString(),
                BackstageConstants.OUTPUT_CLASSPATH_DIR
            )
        )
        logger.info("Created classpath directory at {}", classpathDir.toFile().absolutePath)
        modulesDir = Files.createDirectories(
            Paths.get(
                builder.windowsDir.absolutePathString(),
                BackstageConstants.OUTPUT_MODULES_DIR
            )
        )
        logger.info("Created modules directory at {}", modulesDir.toFile().absolutePath)
    }

    private fun copyLauncher() {
        val launcherPath = Path.of(
            builder.launchCodeDir.absolutePathString(),
            BackstageConstants.OUTPUT_LAUNCHER_DIST_DIR,
            BackstageConstants.LAUNCHER_NAME_WINDOWS
        )

        var exeName = builder.config.deployment.applicationInfo.exeName
        if (!exeName.endsWith(".exe")) {
            exeName = "$exeName.exe"
        }
        logger.debug(
            "Copying Windows launcher from {} to {}", launcherPath.absolutePathString(), builder.windowsDir.resolve(
                exeName
            )
        )
        Files.copy(
            launcherPath,
            builder.windowsDir.resolve(exeName),
            StandardCopyOption.COPY_ATTRIBUTES
        )
    }

    private fun copyWindowsIcons() {
        builder.config.deployment.applicationInfo.windows.icons.forEach {
            val name = if (it.target != null) it.target else it.source
            val target = builder.windowsDir.resolve(name)
            logger.debug("Icon target: {}", target.toFile().absolutePath)
            Files.createDirectories(target.parent)
            builder.config.contentService?.get(it.sha256 ?: throw BackstageException(ErrorInfo.EMPTY_CONTENT_CHECKSUM))
                .use {
                    Files.copy(it, target, StandardCopyOption.REPLACE_EXISTING)
                }
        }
    }

    private fun copySplashScreen() {
        val splash = builder.config.deployment.applicationInfo.splashScreen
        if (splash!= null) {
            val name = if (splash.target != null) splash.target else splash.source
            val target = builder.windowsDir.resolve(name)
            logger.debug("Splash screen target: {}", target.toFile().absolutePath)
            Files.createDirectories(target.parent)
            builder.config.contentService?.get(splash.sha256 ?: throw BackstageException(ErrorInfo.EMPTY_CONTENT_CHECKSUM)).use {
                Files.copy(it, target, StandardCopyOption.REPLACE_EXISTING)
            }
        }
    }
}